transmogrifier/sources/transformer.py

```diff
@@ -55,73 +68,246 @@ def __next__(self) -> TimdexRecord:
+        """
+        Transform a source record into a TIMDEX record.
+
+        Must be overridden by format subclasses.
+
+        Args:
+            source_record: A single source record.
+        """
+        return None
+
+    @classmethod
+    @abstractmethod
+    def get_valid_title(cls, source_record_id: str, source_record: JSON | Tag) -> str:
+        """
+        Retrieves main title(s) from a source record and returns a valid title string.
+
+        Must be overridden by source subclasses.
+
+        Args:
+            source_record_id: Record identifier for the source record.
+            source_record: A single source record.
+        """
+        return ""
+
+    @classmethod
+    @abstractmethod
+    def get_source_link(
+        cls, source_base_url: str, source_record_id: str, source_record: JSON | Tag
+    ) -> str:
+        """
+        Class method to set the source link for the item.
+
+        Must be overridden by source subclasses.
+
+        Args:
+            source_base_url: Source base URL.
+            source_record_id: Record identifier for the source record.
+            source_record: A single source record.
+        """
+        return ""
+
+    @classmethod
+    @abstractmethod
+    def get_timdex_record_id(
+        cls, source: str, source_record_id: str, source_record: Tag
+    ) -> str:
+        """
+        Class method to set the TIMDEX record id.
+
+        Must be overridden by source subclasses.
+
+        Args:
+            source: Source name.
+            source_record_id: Record identifier for the source record.
+            source_record: A single source record.
+        """
+        return ""
+
+
+class XmlTransformer(Transformer):
+    """Base transformer class."""
+
+    @final
+    @classmethod
+    def parse_source_file(cls, input_file: str) -> Iterator[Tag]:
+        """
+        Parse source file and return source records via an iterator.
+
+        May not be overridden.
+        """
+        with open(input_file, "rb") as file:
+            for _, element in etree.iterparse(
+                file,
+                tag="{*}record",
+                encoding="utf-8",
+                recover=True,
+            ):
+                record_string = etree.tostring(element, encoding="utf-8")
+                record = BeautifulSoup(record_string, "xml")
+                yield record
+                element.clear()
+
+    @abstractmethod
+    def get_optional_fields(self, source_record: Tag) -> Optional[dict]:
         """
         Retrieve optional TIMDEX fields from an XML record.
 
         Must be overridden by source subclasses.
 
         Args:
-            xml: A BeautifulSoup Tag representing a single XML record
+            source_record: A BeautifulSoup Tag representing a single XML record
         """
         return {}
 
     @classmethod
     @abstractmethod
-    def get_main_titles(cls, xml: Tag) -> list[Tag]:
+    def get_main_titles(cls, source_record: Tag) -> list[Tag]:
```

Comment:
Arbitrary where to note this, but like the change from `xml` (which is problematic for shadowing a stdlib name and also just kind of generic) to `source_record`.  This will make context switching from XML to JSON easier when looking at multiple transformations.